File: js/action/CCActionMagicSlow.js
;
/*!
 *
 *  SpriteForest JavaScript Library v1.0.0
 *  http://github.com/moky/SpriteForest/js/
 *
 *  Copyright (c) 2014 moKy at slanissue.com
 *  Released under the MIT license
 *  http://en.wikipedia.org/wiki/MIT_License
 *
 */
//
//  CCActionMagicSlow.js
//  SpriteForest-HTML5
//
//  Created by moKy on Mar. 1, 2014.
//

!function(sf) {
	'use strict';
	var cc = sf.cc;
	var cn = sf.cn;
	
	//
	//  constructor
	//
	function ActionMagicSlow() {
		cc.ActionMagic.prototype.ctor.call(this);
	}
	
	//
	//  static functions
	//
	
	function showIndicator() {
		
	}
	
	function hideIndicator() {
		
	}
	
	//
	//  instance functions
	//
	
	//--------------------------------------------------------------------------
	
	var prop = {
		ctor: ActionMagicSlow,
		
		// properties
	};
	
	// extends
	cc.ActionMagicSlow = cc.ActionMagic.extend(prop);
	
	// static
	cc.ActionMagicSlow.create = cc.ActionMagic.create;
	
	cc.ActionMagicSlow.showIndicator = showIndicator;
	cc.ActionMagicSlow.hideIndicator = hideIndicator;
	
}(SpriteForest);

//==============================================================================

!function(sf) {
	'use strict';
	var cc = sf.cc;
	var cn = sf.cn;
	
	function MagicBuilder(name, execute) {
		
		var prop = {
			ctor: cc.ActionMagicSlow.prototype.ctor,
			execute: execute,
		};
		
		var ccClass = cc.ActionMagicSlow.extend(prop);
		ccClass.create = cc.ActionMagicSlow.create;
		ccClass.showIndicator = cc.ActionMagicSlow.showIndicator;
		ccClass.hideIndicator = cc.ActionMagicSlow.hideIndicator;
		
		cc['Action' + name] = ccClass;
		return ccClass;
	}
	
	// LoadingIndicator
	function executeLoadingIndicator() {
		// do nothing
	}
	MagicBuilder('LoadingIndicator', executeLoadingIndicator);
	
	function performForestWithDictonary(dict) {
		var myLord = sf.Lord.getInstance();
		var history = dict['History'];
		if (history) {
			if (history === 'goBack()') {
				if (myLord.goBack()) {
					cn.log('back to previous forest/scene');
				} else {
					cn.log('failed to go back');
				}
			} else {
				cn.log('unrecognized history info');
			}
			return;
		}
		
		var path = dict['Path'];
		var scene = dict['Scene'];
		
		if (path) {
			if (!scene || typeof scene === 'string') {
				scene = dict;
			}
			if (myLord.runIntoForest(path, scene)) {
				cn.log('run into forest ok, path: ' + path + ', scene: ' + scene);
			} else {
				cn.log('invalid forest');
			}
		} else if (scene) {
			if (myLord.performScene(scene)) {
				cn.log('perform scene ok, scene: ' + scene)
			} else {
				cn.log('failed to perform scene');
			}
		} else {
			cn.error(dict);
		}
	}
	
	// Forest
	function executeForest() {
		performForestWithDictonary(this._dict);
	}
	MagicBuilder('Forest', executeForest);
	
	// Scene
	function executeScene() {
		// this is a 'Scene' action info, transform it to a 'Forest' action
		var dict = { 'Scene' : this._dict };
		performForestWithDictonary(dict);
	}
	MagicBuilder('Scene', executeScene);
	
}(SpriteForest);

File: spriteforest-x/action/CCActionMagicSlow.cpp
//
//  CCActionMagicSlow.cpp
//  SpriteForest-X
//
//  Created by Moky on 12-11-28.
//  Copyright 2012 Slanissue Inc. All rights reserved.
//

#include "CCActionMagicSlow.h"

#include "../base/SFLord.h"

NS_SF_BEGIN

USING_NS_CC;

// MagicSlow
void CCActionMagicSlow::showIndicator(void)
{
	CNLog("not implement");
}

void CCActionMagicSlow::hideIndicator(void)
{
	CNLog("not implement");
}

// LoadingIndicator
void CCActionLoadingIndicator::execute(void) const
{
	if (!m_pDict)
	{
		CNLog("could not happen!");
		return;
	}
	CCString * pHidden = (CCString *)m_pDict->objectForKey("Hidden");
	if (!pHidden)
	{
		CNLog("invalid dict");
		return;
	}
	if (pHidden->boolValue())
	{
		hideIndicator();
	}
	else
	{
		showIndicator();
	}
}

// Forest
void CCActionForest_performForestWithDictonary(CCDictionary * pDict)
{
	SFLord * myLord = SFLord::sharedLord();
	if (!myLord)
	{
		CNLog("could not happen!");
		return;
	}
	CCString * pHistory = (CCString *)pDict->objectForKey("History");
	if (pHistory)
	{
		if (strcasecmp(pHistory->getCString(), "goBack()") == 0)
		{
			if (myLord->goBack())
			{
				CNLog("back to previous forest/scene");
			}
			else
			{
				CNLog("failed to go back");
			}
		}
		else
		{
			CNLog("unrecognized history");
		}
		return;
	}
	
	CCString * pPath = (CCString *)pDict->objectForKey("File");
	CCObject * pScene = pDict->objectForKey("Scene");
	
	if (pPath)
	{
		if (!pScene || dynamic_cast<CCString *>(pScene))
		{
			pScene = pDict;
		}
		if (myLord->runIntoForest(pPath->getCString(), pScene))
		{
			CNLog("OK");
		}
		else
		{
			CNLog("invalid forest");
		}
	}
	else if (pScene)
	{
		if (myLord->performScene(pScene))
		{
			CNLog("OK");
		}
		else
		{
			CNLog("failed to perform scene");
		}
	}
	else
	{
		CNLog("error");
	}
}

void CCActionForest_performForestWithDictonary(CCDictionary * pDict, const bool bShowIndicator)
{
	if (bShowIndicator)
	{
		CCActionForest::showIndicator();
		CCActionForest_performForestWithDictonary(pDict);
		CCActionForest::hideIndicator();
	}
	else
	{
		CCActionForest_performForestWithDictonary(pDict);
	}
}

void CCActionForest::execute(void) const
{
	if (!m_pDict)
	{
		CNLog("could not happen!");
		return;
	}
	CCString * pShowIndicator = (CCString *)m_pDict->objectForKey("ShowIndicator");
	bool flag = pShowIndicator ? pShowIndicator->boolValue() : true;
	CCActionForest_performForestWithDictonary(m_pDict, flag);
}

// Scene
void CCActionScene::execute(void) const
{
	if (!m_pDict)
	{
		CNLog("could not happen!");
		return;
	}
	CCString * pShowIndicator = (CCString *)m_pDict->objectForKey("ShowIndicator");
	bool flag = pShowIndicator ? pShowIndicator->boolValue() : true;
	// this is a 'Scene' action info, transform it to a 'Forest' action info
	CCDictionary * pDict = CCDictionary::dictionary();
	pDict->setObject(m_pDict, "Scene");
	CCActionForest_performForestWithDictonary(pDict, flag);
}

NS_SF_END
